tools/chpldoc/findUndocumentedSymbols.py

```diff
@@ -0,0 +1,387 @@
+"""
+
+from collections import defaultdict
+from typing import Generator, List
+import sys
+import os
+import argparse as ap
+import itertools
+
+try:
+    import chapel.core as dyno
+    from chapel import each_matching
+except ModuleNotFoundError:
+    print("chapel python bindings not found")
+    exit(1)
+
+
+"""
+helpers
+"""
+
+
+def get_module(n: dyno.AstNode) -> dyno.AstNode:
+    if n is None:
+        return None
+    if isinstance(n, dyno.Module):
+        return n
+    if parent := n.parent():
+        return get_module(parent)
+    return None
+
+
+def node_or_parent_has_attribute(node: dyno.AstNode, marker: str) -> bool:
+    """
+    a symbol is has an attribute if it has an attribute that matches the argument, or is parent does
+    """
+    if (attrs := node.attribute_group()) and any(a.name() == marker for a in attrs):
+        return True
+
+    if (parent := node.parent()) and node_or_parent_has_attribute(parent, marker):
+        return True
+
+    return False
+
+
+def is_deprecated(node: dyno.AstNode) -> bool:
+    """a node is deprecated if it or its parent has an attribute that is deprecated"""
+    return node_or_parent_has_attribute(node, "deprecated")
+
+
+def is_unstable(node: dyno.AstNode) -> bool:
+    """a node is unstable if it or its parent has an attribute that is unstable"""
+    return node_or_parent_has_attribute(node, "unstable")
+
+
+def is_nodoc(node: dyno.AstNode) -> bool:
+    """
+    a symbol is marked nodoc for the following reasons
+    - symbol has attribute `@chpldoc.nodoc`
+    - symbol name is prefixed with `chpl_` and symbol does not have `pragma "chpldoc ignore chpl prefix"`
+    - symbol is private
+    - symbol is a child of a node that is nodoc
+    """
+    if (attrs := node.attribute_group()) and any(
+        a.name() == "chpldoc.nodoc" for a in attrs
+    ):
+        return True
+
+    if (
+        hasattr(node, "name")
+        and node.name().startswith("chpl_")
+        and "chpldoc ignore chpl prefix" not in node.pragmas()
+    ):
+        return True
+
+    if hasattr(node, "visibility") and node.visibility() == "private":
+        return True
+
+    if (parent := node.parent()) and is_nodoc(parent):
+        return True
+
+    return False
+
+
+def is_docstring_comment(c: dyno.Comment) -> bool:
+    """c is a docstring if it doesn't begin with '//'"""
+    return not c.text().startswith("//")
+
+
+def get_node_name(node: dyno.AstNode) -> str:
+    """
+    get the name of a node, even when it may not have a `name()` method
+
+    if this is a member of a aggregate type, also gets that name
+    if this is a member of a module, also gets that name
+
+    For a multidecl, gets all the names
+
+    """
+
+    def get_simple_node_name(node: dyno.AstNode) -> str:
+        name: str = None
+        if hasattr(node, "name"):
+            name = node.name()
+        elif isinstance(node, dyno.MultiDecl):
+            name = "MultiDecl[" + ", ".join(c.name() for c in node) + "]"
```

Comment:
Ah, so this syntax isn't Chapel/Dyno-driven. I think that you should make this function return a list rather than a string, and have the code that calls it emit multiple warnings, one for each element of the list. That way, you would just warn about `C.a` and `C.b` missing docs, rather than inventing the `C.MultiDecl[a,b]` syntax.